storage/src/main/java/org/apache/kafka/server/log/internals/LogOffsetMetadata.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    public LogOffsetMetadata(long messageOffset,
+                             long segmentBaseOffset,
+                             int relativePositionInSegment) {
+        this.messageOffset = messageOffset;
+        this.segmentBaseOffset = segmentBaseOffset;
+        this.relativePositionInSegment = relativePositionInSegment;
+    }
+
+    public long messageOffset() {
+        return messageOffset;
+    }
+
+    public long segmentBaseOffset() {
+        return segmentBaseOffset;
+    }
+
+    public int relativePositionInSegment() {
+        return relativePositionInSegment;
+    }
+
+    // check if this offset is already on an older segment compared with the given offset
+    public boolean onOlderSegment(LogOffsetMetadata that) {
+        if (messageOffsetOnly())
+            throw new KafkaException(this + " cannot compare its segment info with " + that + " since it only has message offset info");
+
+        return this.segmentBaseOffset < that.segmentBaseOffset;
+    }
+
+    // check if this offset is on the same segment with the given offset
+    private boolean onSameSegment(LogOffsetMetadata that) {
+        if (messageOffsetOnly())
+            throw new KafkaException(this + " cannot compare its segment info with " + that + " since it only has message offset info");
+
+        return this.segmentBaseOffset == that.segmentBaseOffset;
+    }
+
+    // compute the number of bytes between this offset to the given offset
+    // if they are on the same segment and this offset precedes the given offset
+    public int positionDiff(LogOffsetMetadata that) {
+        if (!onSameSegment(that))
+            throw new KafkaException(this + " cannot compare its segment position with " + that + " since they are not on the same segment");
+
+        return this.relativePositionInSegment - that.relativePositionInSegment;
+    }
+
+    // decide if the offset metadata only contains message offset info
+    public boolean messageOffsetOnly() {
+        return segmentBaseOffset == UNIFIED_LOG_UNKNOWN_OFFSET && relativePositionInSegment == UNKNOWN_FILE_POSITION;
+    }
+
+    @Override
+    public String toString() {
+        return "LogOffsetMetadata{" +
+                "messageOffset=" + messageOffset +
+                ", segmentBaseOffset=" + segmentBaseOffset +
+                ", relativePositionInSegment=" + relativePositionInSegment +
+                '}';
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        LogOffsetMetadata that = (LogOffsetMetadata) o;
+        return messageOffset == that.messageOffset && segmentBaseOffset == that.segmentBaseOffset && relativePositionInSegment == that.relativePositionInSegment;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(messageOffset, segmentBaseOffset, relativePositionInSegment);
```

Comment:
This implementation results in boxing all the values and then adding them to an array, etc. To avoid any regressions, we should go with the approach that doesn't do any of that.